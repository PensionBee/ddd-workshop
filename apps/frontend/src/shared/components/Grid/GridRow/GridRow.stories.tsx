import { ThemeProvider } from "styled-components";
import type { ComponentMeta } from "@storybook/react";
import { GlobalStyles } from "@shared/styling/globalStyles";
import { theme } from "@shared/styling/theme";
import { Box } from "../../Box/Box";
import { GridContainer } from "../GridContainer/GridContainer";
import { GridItem } from "../GridItem/GridItem";
import { GridRow } from "./GridRow";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const themeDecorator = (Story: any) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Box position="relative">
        <GridContainer backgroundColor="grey8">
          <Box
            d="flex"
            direction="column"
            backgroundColor="grey6"
            pt={20}
            pb={20}
            rowGap={20}
          >
            <Story />
          </Box>
        </GridContainer>
      </Box>
    </ThemeProvider>
  );
};

export default {
  title: "Shared/Grid/Row",
  component: GridRow,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
  decorators: [themeDecorator],
} as ComponentMeta<typeof GridRow>;

const Child = () => <Box backgroundColor="grey4" w="100%" h={200} />;

export const OneRow = () => (
  <GridRow>
    <Child />
  </GridRow>
);

export const TwoRows = () => {
  const items = [0, 1];
  return (
    <>
      {items.map((item) => (
        <GridRow key={item}>
          <Child />
        </GridRow>
      ))}
    </>
  );
};

export const RowAlignStart = () => (
  <GridRow>
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem.
      </Box>
    </GridItem>
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem, maiores, consectetur sed provident ratione laborum
        autem nihil enim aliquid, ipsum saepe excepturi vero. Veniam perferendis
        tempora repellat? Molestiae quaerat autem illum enim consequuntur
        repellat alias quam iure, aspernatur eum?
      </Box>
    </GridItem>
  </GridRow>
);

export const RowAlignCenter = () => (
  <GridRow align="center">
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem.
      </Box>
    </GridItem>
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem, maiores, consectetur sed provident ratione laborum
        autem nihil enim aliquid, ipsum saepe excepturi vero. Veniam perferendis
        tempora repellat? Molestiae quaerat autem illum enim consequuntur
        repellat alias quam iure, aspernatur eum?
      </Box>
    </GridItem>
  </GridRow>
);

export const RowAlignEnd = () => (
  <GridRow align="end">
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem.
      </Box>
    </GridItem>
    <GridItem span={6}>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem, maiores, consectetur sed provident ratione laborum
        autem nihil enim aliquid, ipsum saepe excepturi vero. Veniam perferendis
        tempora repellat? Molestiae quaerat autem illum enim consequuntur
        repellat alias quam iure, aspernatur eum?
      </Box>
    </GridItem>
  </GridRow>
);

export const RowGap = () => (
  <GridRow rowGap={{ xs: 20, md: 28, lg: 36 }}>
    <GridItem>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem, maiores, consectetur sed provident ratione laborum
        autem nihil enim aliquid, ipsum saepe excepturi vero. Veniam perferendis
        tempora repellat? Molestiae quaerat autem illum enim consequuntur
        repellat alias quam iure, aspernatur eum?
      </Box>
    </GridItem>
    <GridItem>
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet quibusdam
        cupiditate dolorem, maiores, consectetur sed provident ratione laborum
        autem nihil enim aliquid, ipsum saepe excepturi vero. Veniam perferendis
        tempora repellat? Molestiae quaerat autem illum enim consequuntur
        repellat alias quam iure, aspernatur eum?
      </Box>
    </GridItem>
  </GridRow>
);

export const RowCentered = () => (
  <GridRow>
    <GridItem align="center">
      <Box backgroundColor="grey4">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit.
      </Box>
    </GridItem>
  </GridRow>
);
