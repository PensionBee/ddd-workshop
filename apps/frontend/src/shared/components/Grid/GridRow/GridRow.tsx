import { Box } from "../../Box/Box";

export type GridRowProps = {
  children: React.ComponentProps<typeof Box>["children"];
  /** Element to render as */
  as?: React.ComponentProps<typeof Box>["as"];
  /** Vertical alignment */
  align?: React.ComponentProps<typeof Box>["align"];
  /** Border bottom in px */
  borderBottom?: React.ComponentProps<typeof Box>["borderBottom"];
  /** Border colour */
  borderColor?: React.ComponentProps<typeof Box>["borderColor"];
  /** Border top in px */
  borderTop?: React.ComponentProps<typeof Box>["borderTop"];
  /** If wrap should be disabled */
  nowrap?: React.ComponentProps<typeof Box>["nowrap"];
  /** Padding bottom in px */
  pb?: React.ComponentProps<typeof Box>["pb"];
  /** Padding top in px */
  pt?: React.ComponentProps<typeof Box>["pt"];
  /** Horizontal space between elements */
  rowGap?: React.ComponentProps<typeof Box>["rowGap"];
  /** Background color */
  backgroundColor?: React.ComponentProps<typeof Box>["backgroundColor"];
};

export const GridRow: React.FC<GridRowProps> = ({
  children,
  as,
  align,
  borderBottom,
  borderColor,
  borderTop,
  nowrap,
  pb,
  pt,
  rowGap,
  backgroundColor,
  ...props
}) => {
  return (
    <Box
      data-rc="GridRow"
      as={as}
      d="flex"
      align={align}
      borderBottom={borderBottom}
      borderColor={borderColor}
      borderTop={borderTop}
      mx={{ xs: -12, md: -14, lg: -24 }}
      nowrap={nowrap}
      pb={pb}
      pt={pt}
      rowGap={rowGap}
      backgroundColor={backgroundColor}
      w={{
        xs: "calc(100% + 24px)",
        md: "calc(100% + 28px)",
        lg: "calc(100% + 48px)",
      }}
      {...props}
    >
      {children}
    </Box>
  );
};
