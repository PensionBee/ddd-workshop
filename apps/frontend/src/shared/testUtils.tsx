import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
