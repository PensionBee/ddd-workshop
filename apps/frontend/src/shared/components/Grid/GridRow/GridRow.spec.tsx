import { render, vi } from "@shared/testUtils";
import { GridRow } from "./GridRow";

describe("GridRow", () => {
  const rowContent = "GridRow Content";
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(<GridRow>{rowContent}</GridRow>);

      expect(getByText(rowContent)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render as elements", () => {
      const { container, asFragment } = render(
        <GridRow as="span">{rowContent}</GridRow>
      );

      expect(container.getElementsByTagName("span").length).toBe(1);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
