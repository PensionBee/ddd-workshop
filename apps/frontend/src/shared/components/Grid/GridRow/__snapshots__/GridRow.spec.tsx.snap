// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GridRow > Default > should render as elements 1`] = `
<DocumentFragment>
  <span
    class="sc-bczSft fnWBbu"
    data-rc="GridRow"
  >
    GridRow Content
  </span>
</DocumentFragment>
`;

exports[`GridRow > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft fnWBbu"
    data-rc="GridRow"
  >
    GridRow Content
  </div>
</DocumentFragment>
`;

exports[`GridRow Default should render as elements 1`] = `
<DocumentFragment>
  <span
    class="sc-bcXHqe cCkuwp"
    data-rc="GridRow"
  >
    GridRow Content
  </span>
</DocumentFragment>
`;

exports[`GridRow Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe cCkuwp"
    data-rc="GridRow"
  >
    GridRow Content
  </div>
</DocumentFragment>
`;
